test/e2e/report_dynamic_input_data_test.go

```diff
@@ -163,3 +169,129 @@ func testReportsProduceData(t *testing.T, testReportingFramework *reportingframe
+			if foundOnce && strings.Contains(err.Error(), "not found") {
+				return true, errors.New("report was deleted after being created")
+			}
+		}
+		// if we're exiting due to timeout, it's a fail
+		return false, nil
+	})
+	if err != nil {
+		assert.Contains(t, err.Error(), "was deleted", "report should have been deleted by retention period being reached")
+	}
+}
+
+func testReportIsNotDeletedWhenReportDependsOnIt(t *testing.T, testReportingFramework *reportingframework.ReportingFramework) {
+	testQueryName := "namespace-cpu-usage"
+
+	// cron schedule to run every minute
+	cronSchedule := &metering.ReportSchedule{
+		Period: metering.ReportPeriodCron,
+		Cron: &metering.ReportScheduleCron{
+			Expression: "*/1 * * * *",
+		},
+	}
+	var foundOnce = false
+	expiration := &metav1.Duration{Duration: 30.0 * time.Second}
+	// If a subReport is in dependency unmet state it won't be deleted and it can take awhile for import to happen.
+	// 4 minutes seems to be enough time for that setup, and the delete, to reliably happen.
+	waitTimeWhileCheckDeletion := 4 * time.Minute
+
+	subReport := testReportingFramework.NewSimpleReport(
+		"subreport-should-not-delete",
+		testQueryName,
+		cronSchedule,
+		nil,
+		nil,
+		expiration,
+	)
+	t.Logf("creating subReport %s and waiting %s to finish", subReport.Name, waitTimeWhileCheckDeletion)
+
+	createErr := testReportingFramework.CreateMeteringReport(subReport)
+	if createErr != nil {
+		assert.Failf(t, "Creating subReport should not have an error", createErr.Error())
+	}
+
+	newDefault := func(s string) *json.RawMessage {
+		v := json.RawMessage(s)
+		return &v
+	}
+
+	report := testhelpers.NewReport(
+		"report-depends-on-subreport",
+		testReportingFramework.Namespace,
+		testQueryName,
+		[]metering.ReportQueryInputValue{
+			metering.ReportQueryInputValue{
+				Name: "Report",
+				Value: newDefault(`"` + "subreport-should-not-delete" + `"`),
+			},
+		},
+		nil,
+		nil,
+		metering.ReportStatus{},
+		cronSchedule,
+		false,
+		nil,
+	)
+
+	createErr = testReportingFramework.CreateMeteringReport(report)
+	if createErr != nil {
+		assert.Failf(t, "Creating report should not have an error", createErr.Error())
+	}
+
+	err := wait.Poll(time.Second*5, waitTimeWhileCheckDeletion, func() (bool, error) {
+		_, err := testReportingFramework.GetMeteringReport("subReport-should-not-delete")
+		// `foundOnce` provides a latch that allows us to wait for subReport creation then exit on deletion
+		if err == nil {
+			foundOnce = true
+		}
+		if err != nil {
+			if foundOnce && strings.Contains(err.Error(), "not found") {
+				return true, errors.New("subReport was deleted after being created")
+			}
+		}
+		// if we're exiting due to timeout, it's correct, because the subReport should not be deleted
+		return false, nil
+	})
+	assert.True(t, strings.Contains(
+		err.Error(),
+		"timed out waiting"),
+		"should have timed out waiting on subReport expiration as subReport should not have been deleted",
+	)
```

Comment:
We may want to also validate that an event got created with the `ExpiredReportHasDependencies` warning type.

Reply:
We can also use [`assert.Contains`](https://godoc.org/github.com/stretchr/testify/assert#Contains) here instead of doing a blanket `assert.True` that's difficult to read.

Reply:
ah I like a typed check better than the strings checking anyway.

Reply:
done in latest commit. though I couldn't quickly work out the filtering. I don't think there are many events to loop on in test so not a big deal, IMO